# Проект по SQL

# Описание проекта

Коронавирус застал мир врасплох, изменив привычный порядок вещей. На какое-то время жители городов перестали выходить на улицу, посещать кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.
Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

# Описание данных

**Таблица `books`**

Содержит данные о книгах:

- book_id — идентификатор книги;
- author_id — идентификатор автора;
- title — название книги;
- num_pages — количество страниц;
- publication_date — дата публикации книги;
- publisher_id — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- author_id — идентификатор автора;
- author — имя автора.
- Таблица publishers
- Содержит данные об издательствах:
- publisher_id — идентификатор издательства;
- publisher — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- rating_id — идентификатор оценки;
- book_id — идентификатор книги;
- username — имя пользователя, оставившего оценку;
- rating — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах:

- review_id — идентификатор обзора;
- book_id — идентификатор книги;
- username — имя автора обзора;
- text — текст обзора.

# Схема данных

![Схема данных](https://pictures.s3.yandex.net/resources/scheme_1589269096.png)

# Ход выполнения проекта

**Шаг 1. Загрузить данные и вывести первые строки**

**Шаг 2. Задание 1**
- Посчитайте, сколько книг вышло после 1 января 2000 года

**Шаг 3. Задание 2**
- Для каждой книги посчитайте количество обзоров и среднюю оценку

**Шаг 4. Задание 3**
- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры

**Шаг 5. Задание 4**
- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

**Шаг 6. Задание 5**
- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок

**Шаг 7. Общий вывод**

### Загрузить данные и вывести первые строки

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine

Для удобства напишем функцию, которая будет выводить первые строки таблиц.

In [4]:
# напишем функция для вывода первых 5 строк датафрейма
def first_lines(table):
    # сформируем запрос
    query = ''' 
    SELECT * 
    FROM {}
    LIMIT 5
    '''.format(table) 
    
    con=engine.connect()
    request = pd.io.sql.read_sql(sql=text(query), con = con)
    
    # выведем на экран
    print('Таблица', table)
    display(request)

In [5]:
# выведем первые строки таблицы books
first_lines('books')

Таблица books


book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

Таблица содержит данные об идентификаторе книги, идентификаторе автора, название книги, количество страниц, дата публикации книги и идентификатор издателя.

In [6]:
# выведем первые строки таблицы authors
first_lines('authors')

Таблица authors


author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

Таблица содержит данные об идентификаторе автора и имя автора.

In [7]:
# выведем первые строки таблицы publishers
first_lines('publishers')

Таблица publishers


publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

Таблица содержит данные об идентификаторе издательства и название издательства.

In [8]:
# выведем первые строки таблицы ratings
first_lines('ratings')

Таблица ratings


rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

Таблица содержит данные об идентификаторе оценки; идентификатор книги; имя пользователя, оставившего оценку и оценка книги.

In [9]:
# выведем первые строки таблицы reviews
first_lines('reviews')

Таблица reviews


review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

Таблица содержит данные об идентификаторе обзора, идентификатор книги, имя автора обзора, текст обзора.

### Задание №1

- Посчитайте, сколько книг вышло после 1 января 2000 года

In [10]:
# сформируем запрос
query_1 = ''' SELECT COUNT(*)
            FROM books WHERE publication_date > '2000-01-01'
        '''
request_1 = pd.io.sql.read_sql(sql=text(query_1), con = engine.connect())

# выведем на экран
print(f'После 1 января 2000 года вышло {request_1.iloc[0,0]} книг')

После 1 января 2000 года вышло 819 книг


После 1 января 2000 года вышло 819 книг.

### Задание 2

- Для каждой книги посчитайте количество обзоров и среднюю оценку

In [11]:
# сформируем запрос
query_2 = '''SELECT b.book_id,
                    b.author_id,
                    b.title,
                    COUNT(DISTINCT re.review_id) AS reviews,
                    ROUND(AVG(ra.rating), 2) AS avg_rating                   
            FROM books AS b
            LEFT JOIN reviews AS re ON b.book_id = re.book_id
            LEFT JOIN ratings AS ra ON b.book_id = ra.book_id
            GROUP BY b.title, b.book_id
            ORDER BY reviews DESC'''

request_2 = pd.io.sql.read_sql(sql=text(query_2), con = engine.connect())

# выведем на экран
display(request_2)

book_id  author_id                                              title  \
0        948        554                            Twilight (Twilight  #1)   
1        963        527                                Water for Elephants   
2        734        278                                   The Glass Castle   
3        302        236  Harry Potter and the Prisoner of Azkaban (Harr...   
4        695        395  The Curious Incident of the Dog in the Night-Time   
..       ...        ...                                                ...   
995       83        174    Anne Rice's The Vampire Lestat: A Graphic Novel   
996      808        350                            The Natural Way to Draw   
997      672        144   The Cat in the Hat and Other Dr. Seuss Favorites   
998      221        151                          Essential Tales and Poems   
999      191        568  Disney's Beauty and the Beast (A Little Golden...   

     reviews  avg_rating  
0          7        3.66  
1          6        3.98  
2          6        4.21  
3          6        4.41  
4          6        4.08  
..       ...         ...  
995        0        3.67  
996        0        3.00  
997        0        5.00  
998        0        4.00  
999        0        4.00  

[1000 rows x 5 columns]

Книгой, с наибольшим количеством обзоров, оказалась - Twilight (Twilight #1) - 7 обзоров и средней оценкой 3.66.

### Задание 3

- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры

In [12]:
# сформируем запрос
query_3 = '''
            SELECT  p.publisher,
                    COUNT(b.book_id) AS count_books
            FROM publishers AS p
            LEFT JOIN books AS b ON p.publisher_id = b.publisher_id
            WHERE b.num_pages > 50
            GROUP BY p.publisher
            ORDER BY count_books DESC
            LIMIT 1
        '''
request_3 = pd.io.sql.read_sql(sql=text(query_3), con = engine.connect())

# выведем на экран
display(request_3)

publisher  count_books
0  Penguin Books           42

Издательство, выпустившее наибольшее число книг толще 50 страниц - Penguin Books (42 книги).

### Задание 4

- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [13]:
# сформируем запрос
query_4 = '''
            SELECT a.author AS author_name,
                    AVG(r.rating) AS avg_rating
            FROM authors AS a
            LEFT JOIN books AS b ON a.author_id = b.author_id
            LEFT JOIN ratings AS r ON b.book_id = r.book_id
            WHERE b.book_id IN (SELECT book_id
                                FROM ratings
                                GROUP BY book_id
                                HAVING COUNT(rating) >=50)
            GROUP BY author_name  
            ORDER BY avg_rating DESC
            LIMIT 1
                                
        '''
request_4 = pd.io.sql.read_sql(sql=text(query_4), con = engine.connect())

# выведем на экран
display(request_4)

author_name  avg_rating
0  J.K. Rowling/Mary GrandPré    4.287097

Автор с самой высокой средней оценкой книг (с 50 и более оценками) - J.K. Rowling/Mary GrandPré. Средняя оценка книг - 4.287.

### Задание 5

- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок

In [14]:
# сформируем запрос
query_5 = '''
                SELECT AVG(t_1.count_reviews) AS avg_reviews
                FROM (SELECT    username,
                                COUNT(DISTINCT review_id) AS count_reviews
                        FROM reviews
                        WHERE username IN (
                                        SELECT username
                                        FROM ratings
                                        GROUP BY username
                                        HAVING COUNT(rating) > 48
                                        ) 
                        GROUP BY username
                ) AS t_1
        '''
request_5 = pd.io.sql.read_sql(sql=text(query_5), con = engine.connect())

# выведем на экран
print(f'Среднее количество обзоров от пользователей, которые поставили больше 48 оценок: {request_5.iloc[0,0]}')

Среднее количество обзоров от пользователей, которые поставили больше 48 оценок: 24.0


Среднее количество обзоров от пользователей, которые поставили больше 48 оценок : 24.0

### Общий вывод

В результате анализа базы данных выявлено:
- После 1 января 2000 года вышло 819 книг
- Книгой, с наибольшим количеством обзоров, оказалась - Twilight (Twilight #1) - 7 обзоров и средней оценкой 3.66
- Издательство, выпустившее наибольшее число книг толще 50 страниц - Penguin Books (42 книги)
- Автор с самой высокой средней оценкой книг (с 50 и более оценками) - J.K. Rowling/Mary GrandPré. Средняя оценка книг - 4.287
- Среднее количество обзоров от пользователей, которые поставили больше 48 оценок : 24.0